ed25519/shamir.go

```diff
@@ -1 +1,183 @@
+// caller. This is defined as multiplying the share value by the scale, and
+// leaving the index unchanged.
+func (s *Share) Scale(other *Share, scale *Scalar) {
+	s.Index = other.Index
+	s.Value.Mul(&other.Value, scale)
+}
+
+// func main(){
+// 	var x *edwards25519.Scalar
+// 	x = edwards25519.NewScalar()
+// 	fmt.Println(x.Bytes())
+// }
+// ShareSecret creates Shamir shares for the given secret at the given
+// threshold, and stores them in the given destination slice. In the returned
+// Shares, there will be one share for each index in the indices that were used
+// to construct the Sharer. If k is larger than the number of indices, in which
+// case it would be impossible to reconstruct the secret, an error is returned.
+//
+// Panics: This function will panic if the destination shares slice has a
+// capacity less than n (the number of indices).
+func ShareSecret(dst *Shares, indices []Scalar, secret Scalar, k int) error {
+	coeffs := make([]Scalar, k)
+	return ShareAndGetCoeffs(dst, coeffs, indices, secret, k)
+}
+
+// ShareAndGetCoeffs is the same as ShareSecret, but uses the provided slice to
+// store the generated coefficients of the sharing polynomial. If this function
+// successfully returns, this slice will contain the coefficients of the
+// sharing polynomial, where index 0 is the constant term.
+//
+// Panics: This function will panic if the destination shares slice has a
+// capacity less than n (the number of indices) or the coefficients slice has
+// length less than k, or any of the given indices is the zero element.
+func ShareAndGetCoeffs(dst *Shares, coeffs, indices []Scalar, secret Scalar, k int) error {
+	for _, index := range indices {
+		if index.IsZero() {
+			panic("cannot create share for index zero")
+		}
+	}
+	if k > len(indices) {
+		return fmt.Errorf(
+			"reconstruction threshold too large: expected k <= %v, got k = %v",
+			len(indices), k,
+		)
+	}
+	setRandomCoeffs(coeffs, secret, k)
+
+	// Set shares
+	// NOTE: This panics if the destination slice does not have the required
+	// capacity.
+	*dst = (*dst)[:len(indices)]
+	var eval Scalar
+	for i, ind := range indices {
+		polyEval(&eval, &ind, coeffs)
+		(*dst)[i].Index = ind
+		(*dst)[i].Value = eval
+	}
+
+	return nil
+}
+
+// Sets the coefficients of the Sharer to represent a random degree k-1
+// polynomial with constant term equal to the given secret.
+//
+// Panics: This function will panic if k is greater than len(coeffs).
+func setRandomCoeffs(coeffs []Scalar, secret Scalar, k int) {
+	coeffs = coeffs[:k]
+	coeffs[0] = secret
+
+	// NOTE: If k > len(coeffs), then this will panic when i > len(coeffs).
+	for i := 1; i < k; i++ {
+		coeffs[i] = RandomScalar() // to be written
+	}
+}
+
+// Evaluates the polynomial defined by the given coefficients at the point x
+// and stores the result in y. Modifies y, but leaves x and coeffs unchanged.
+// Normalizes y, so this this is not neccesary to do manually after calling
+// this function.
```

Comment:
Normalisation is only relevent for secp256k1 types due to their internal representation.
```suggestion
// Evaluates the polynomial defined by the given coefficients at the point x
// and stores the result in y. Modifies y, but leaves x and coeffs unchanged.
```